src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs

```diff
@@ -50,20 +58,33 @@ public RoamingVisualStudioProfileOptionPersister(IThreadingContext threadingCont
         {
             Contract.ThrowIfNull(globalOptionService);
 
-            _settingManager = (ISettingsManager)serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
             _globalOptionService = globalOptionService;
 
-            // While the settings persistence service should be available in all SKUs it is possible an ISO shell author has undefined the
-            // contributing package. In that case persistence of settings won't work (we don't bother with a backup solution for persistence
-            // as the scenario seems exceedingly unlikely), but we shouldn't crash the IDE.
-            if (_settingManager != null)
-            {
-                var settingsSubset = _settingManager.GetSubset("*");
-                settingsSubset.SettingChangedAsync += OnSettingChangedAsync;
-            }
+            var serviceContainer = serviceProvider.GetService<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+            _serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+            _serviceBroker.AvailabilityChanged += Sb_AvailabilityChanged;
+
+            ConnectSettingsManager();
+        }
+
+        private void ConnectSettingsManager()
+        {
+            ThreadingContext.JoinableTaskFactory.Run(
+                async () =>
+                {
+                    _brokeredSettingsManager = await _serviceBroker.GetProxyAsync<IBrokeredSettingsManager>(VisualStudioServices.VS2019_4.SettingsManager);
+                    await _brokeredSettingsManager.RequestChangeEventsAsync("*", CancellationToken.None).ConfigureAwait(true);
+                    _brokeredSettingsManager.SettingChanged += OnSettingChanged;
+                    return;
+                });
+        }
+
+        private void Sb_AvailabilityChanged(object sender, ServiceHub.Framework.BrokeredServicesChangedEventArgs e)
+        {
+            ConnectSettingsManager();
```

Comment:
This actually shouldn't be necessary since the IBrokeredSettingsManager implementation runs locally and refreshes its own internal remote connection on AvailabilityChanged.  But if you want to do it on principle, to avoid taking an implicit dependency on the local-ness of IBrokeredSettingsManager, I'm not totally against it.

Reply:
At present, properly switching a proxy from one service implementation to another when you're also wiring up event handlers is complicated to get right. So if Josh is taking that on in his service and guaranteeing the service is always local, that's nice. But we've never publicly stated that even local services are "always available" and in fact I think the opposite is true. Even local services might be hosted by a ServiceHub process, and if that process crashes a client should be prepared to re-query for it. Another possibility is a local service is hosted by an 'extension' that runs in ServiceHub and if we update/remove that extension while VS is running, the AvailabilityChanged will signify when that service is missing or available again. If the settings manager is in proc today, we might want to "push it out" to ServiceHub later. 

So in general, I think we should probably keep advising people to write clients fully capable of re-querying for services according to this.
But by the same token, they should be prepared for the service to *not* be available and have a backup plan (that isn't crashing, preferably).

Now to this case particularly, the settings manager is probably one of those services that people will realistically have a hard time accommodating being missing, and it likely will *not* be 'upgraded' while VS is running. So if you want to query for it once and expect the instance never changes (given Josh's offer that it will always be local), that's probably ok.

Reply:
@AArnott is it even guaranteed that the IBrokeredSettingsManager client would get an AvailabilityChanged event when the server connection changes, given that IBrokeredSettingsManager is proffered locally and only has an internal dependency on a remote service?  I thought I remembered something about AvailabilityChanged being raised only for services you've actually queried, so could that mean that since IBrokeredSettingsManager's availability isn't changing, the client wouldn't get the event?

Reply:
The event is only raised once after you query for it (if you query for it again, it will be raised again), when its availability changes. Its availability changes when it is proffered or unproffered, or when a remote connection is made or dropped over which that service would be acquired.
Since your service is local, the only time the event would be raised is if your package proffered  it (after your package init was complete) or unproffered it later.

Reply:
So it sounds like my service *must* take responsibility for refreshing its internal remote connection on AvailabilityChanged; I can't push that responsibility onto my clients even if I wanted to.

Reply:
I think you're right. If your service directly consumes a service that may come from various sources, you have to build in handling of that source changing.